// models/models.js
var mysql = require('mysql');
var config = require('config');

function Models() {

    this.pool = mysql.createPool({
        host: config.db.host,
        user: config.db.user,
        password: config.db.password,
        database: config.db.name,
        multipleStatements: config.db.multipleStatements
    });

}

Models._query = function(sql, args, callback) {

    if (typeof args === 'function') {
        callback = args;
        args = [];
    }

    return function(err, connection) {

        if (err) {
            return callback(err);
        }

        var finisher = function(err, result) {
            connection.release();
            callback(err, result);
        };

        connection.query(sql, args, finisher);
    }
}

Models.prototype.query = function() {
    var _this = this;
    _this.getConnection(Models._query.apply(_this, arguments));
}

Models.prototype.getConnection = function(callback) {

    this.pool.getConnection(function(err, conn) {
        callback(err, conn);
    });

}

Models.prototype.queryOnConnection = function(args, count, method) {

    var _this = this;
    var args = [].slice.call(args);

    if (args.length == count) {

        method.apply(_this, args);

    } else {

        _this.getConnection(function(err, connection) {

            var callback = args[args.length - 1];

            if (err) {
                return callback(err);
            }

            args[args.length - 1] = function(err) {

                if (err) {
                    connection.rollback();
                    connection.release();
                    return callback.apply(this, arguments);
                }

                connection.commit();
                connection.release();
                return callback.apply(this, arguments);

            };

            args.unshift(connection);
            connection.beginTransaction();
            method.apply(_this, args);

        });

    }

};

module.exports = new Models();
